File: assignments/PPT/DSA/Assignments/Assignment_3/Q7.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <iterator>

using namespace std;

/*
You are given an inclusive range [lower, upper] and a sorted unique integer array
nums, where all elements are within the inclusive range.

A number x is considered missing if x is in the range [lower, upper] and x is not in
nums.

Return the shortest sorted list of ranges that exactly covers all the missing
numbers. That is, no element of nums is included in any of the ranges, and each
missing number is covered by one of the ranges.

Example 1:
Input: nums = [0,1,3,50,75], lower = 0, upper = 99
Output: [[2,2],[4,49],[51,74],[76,99]]

Explanation: The ranges are:
[2,2]
[4,49]
[51,74]
[76,99]

*/



vector<vector<int> > solve(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        vector<vector<int> > rv;
        if (n == 0) {
            int ar[] = {lower, upper};
            rv.push_back(vector<int>(ar,ar+2));
            return rv;
        }
        if (lower < nums[0]) {
            int ar[] = {lower, nums[0] - 1};
            rv.push_back(vector<int>(ar,ar+2));
        }
        for (int i = 0; i < n - 1; i++) {
            if (nums[i + 1] - nums[i] <= 1) {
                continue;
            }
            int ar[] = {nums[i] + 1, nums[i + 1] - 1};
            rv.push_back(vector<int>(ar,ar+2));
        }
        if (upper > nums[n - 1]) {
            int ar[] = {nums[n - 1] + 1, upper};
            rv.push_back(vector<int>(ar,ar+2));
        }

        return rv;
    }


int main(){
    // Remove the following three lines when running in terminal with input or some interface
    #ifndef ONLINE_JUDGE
        freopen("input7.txt","r",stdin);
    #endif

    int t,n,tcn = 0;
    cin>>t;
    while(t--){
        tcn++;
        cin>>n;
        vector<int> ar(n);
        for(int i = 0;i<n;i++){
            cin>>ar[i];
        }
        int l = 0, h = 99;

        vector<vector<int> > ans = solve(ar,l,h);
        for(auto v: ans){
            for(int i : v){
                cout<<i<<" ";
            }
            cout<<endl;
        }
        cout<< "test case "<<tcn<<"done!"<<endl;
    }

    return 0;
}

File: assignments/PPT/DSA/Assignments/Assignment_5/Q7.py
'''
Suppose an array of length n sorted in ascending order is **rotated** between 1 and n times. For example, the array nums = [0,1,2,4,5,6,7] might become:

- [4,5,6,7,0,1,2] if it was rotated 4 times.
- [0,1,2,4,5,6,7] if it was rotated 7 times.

Notice that **rotating** an array [a[0], a[1], a[2], ..., a[n-1]] 1 time results in the array [a[n-1], a[0], a[1], a[2], ..., a[n-2]].

Given the sorted rotated array nums of **unique** elements, return *the minimum element of this array*.

You must write an algorithm that runs in O(log n) time.

**Example 1:**

**Input:** nums = [3,4,5,1,2]

**Output:** 1

**Explanation:**

The original array was [1,2,3,4,5] rotated 3 times.
'''

def findMin(nums: list[int]) -> int:
    l = 0
    h = len(nums) - 1
    if h == 0:
        return nums[0]
    if nums[h] > nums[0]: return nums[0]

    while(l<h):
        m = l + (h-l)//2
        if nums[m] > nums[m+1]:
            return nums[m+1]
        
        if nums[m-1] > nums[m]:
            return nums[m]
        
        if nums[m] > nums[0]:
            l = m+1
        else:
            h = m-1
            
